stellar/contracts/wormhole-contract/src/vaa/mod.rs

```diff
@@ -0,0 +1,216 @@
+    pub nonce: u32,
+    pub emitter_chain: u32,
+    pub emitter_address: BytesN<32>,
+    pub sequence: u64,
+    pub consistency_level: u32,
+    pub payload: Bytes,
+}
+
+impl<'a> TryFrom<(&'a Env, &'a Bytes)> for VAA {
+    type Error = Error;
+
+    fn try_from(value: (&'a Env, &'a Bytes)) -> Result<Self, Self::Error> {
+        let (env, vaa_bytes) = value;
+
+        if vaa_bytes.len() < 6 {
+            return Err(Error::InvalidVAAFormat);
+        }
+
+        let mut reader = BytesReader::new(vaa_bytes);
+
+        let version = u32::from(reader.read_u8()?);
+        let guardian_set_index = reader.read_u32_be()?;
+        let num_signatures = u32::from(reader.read_u8()?);
+
+        let mut signatures = Vec::new(env);
+        for _ in 0..num_signatures {
+            let sig = Signature::parse(env, &mut reader)?;
+            signatures.push_back(sig);
+        }
+
+        let timestamp = reader.read_u32_be()?;
+        let nonce = reader.read_u32_be()?;
+        let emitter_chain = u32::from(reader.read_u16_be()?);
+        let emitter_address = reader.read_bytes_n::<32>(env)?;
+        let sequence = reader.read_u64_be()?;
+        let consistency_level = u32::from(reader.read_u8()?);
+        let payload = reader.remaining_bytes();
+
+        Ok(VAA {
+            version,
+            guardian_set_index,
+            signatures,
+            timestamp,
+            nonce,
+            emitter_chain,
+            emitter_address,
+            sequence,
+            consistency_level,
+            payload,
+        })
+    }
+}
+
+impl VAA {
+
+    /// Serialize the VAA body for hashing
+    #[allow(clippy::arithmetic_side_effects, clippy::cast_possible_truncation)]
+    pub(crate) fn serialize_body(&self, env: &Env) -> Bytes {
+        let mut bytes = Bytes::new(env);
+
+        for i in (0..4).rev() {
+            bytes.push_back(((self.timestamp >> (i * 8)) & 0xFF) as u8);
+        }
+
+        for i in (0..4).rev() {
+            bytes.push_back(((self.nonce >> (i * 8)) & 0xFF) as u8);
+        }
+
+        bytes.push_back((self.emitter_chain >> 8) as u8);
+        bytes.push_back((self.emitter_chain & 0xFF) as u8);
+
+        for byte in self.emitter_address.to_array().iter() {
+            bytes.push_back(*byte);
+        }
+
+        for i in (0..8).rev() {
+            bytes.push_back(((self.sequence >> (i * 8)) & 0xFF) as u8);
+        }
+
+        bytes.push_back(self.consistency_level as u8);
+
+        bytes.append(&self.payload);
+
+        bytes
+    }
+
+    /// Get the body bytes for hashing (everything after signatures)
+    pub(crate) fn get_body_bytes(vaa_bytes: &Bytes) -> Result<Bytes, Error> {
+        if vaa_bytes.len() < 6 {
+            return Err(Error::InvalidVAAFormat);
+        }
+
+        let num_sigs = u32::from(vaa_bytes.get(5).ok_or(Error::InvalidVAAFormat)?);
+        let body_offset = 6u32.saturating_add(66u32.saturating_mul(num_sigs));
+
+        if body_offset >= vaa_bytes.len() {
```

Comment:
Is it possible to have an empty payload? `TryFrom` allows it, but this line doesn't. If it's the case, change in `>` instead of `>=`